Set INITDATA for non-confidential peer pods

snir911 opened the pull request:
Follow up to this change:
https://github.com/openshift/cloud-api-adaptor/pull/218

snir911 committed:
Set INITDATA for non-confidential peer pods

When confidential computing is disabled, set INITDATA in the peer pods
configmap to override the default restrictive CoCo agent policy. This
allows non-confidential workloads to function properly.
When confidential computing is enabled, clear INITDATA in case it
containes the default valued used for the non-coco case.

Signed-off-by: Snir Schreiber <ssheribe@redhat.com>

bpradipt commented on controllers/confidential_handler.go:
```diff
@@ -28,6 +28,10 @@ const (
+
+	// INITDATA value for non-confidential peer pods, this is required in order
+	// to override the default restrictive CoCo agent policy
+	defaultInitdata = "H4sIAAAAAAAAA42UTU/DMAyG7/0VVS+7TUNwQEg7jG3AJMaqdmNH5LVeG5HGJXHGyq8npRLiQNodcrGfOH79EZAFacFlFU7DyJRwfXsTBSfURpBqTZPx1XgSBUEQ1SRF1ow1FhQ5z2g0CmrI3qHA0B3Fbx0QBDkewUoOZ3k+S+IXFEV5IG0S/LBoOLybhqwt/sXST6h97rkkgynnQnkJqpsHIdHr1wiMc1IMQqHux1JQ+YHOPmjhjJqaAWp5xizWlKHxqn5EXiNrkfURm816eXKl9SKtdYFOl/SGeRaGV4pRH8Hl00clZNlPrLF6Iq6lLe4bp+3gLfdGSVfmebxzN3xMDNYMdyS2Uq4qN1w+IEHIU3Z98z6UYEWn4Zc ...
```
May be a reference to the plaintext initdata in source tree will be good for some dev looking at in future ?

gkurz replied:
Since this for non-confidential peer pods only, what about naming the constant `permissiveInitData` or something like that ?

gkurz commented on controllers/confidential_handler.go:
```diff
@@ -94,8 +98,12 @@ func (r *KataConfigOpenShiftReconciler) handleConfidentialPeerPods(state Feature
 	var peerpodsCMData map[string]string
 	if state == Enabled {
 		peerpodsCMData = map[string]string{"DISABLECVM": "false"}
+		// TODO: make sure we clear INITDATA value if not default
```
IIUC the discussion I had with @snir911 on slack, the issue is that if KataConfig is created first, the other branch of the `if` will add the unsecure `defaultInitdata` to the CM and we don't want that to creep into the podvm when the confidential FG is enabled later on, but at the same time we don't want to clear out an INITDATA provided by the user. Right ?

This is an ownership problem. If the controller is allowed to set the initdata, it means it owns it and it will always take precedence on whatever the user puts there. The user provided initdata should come from some other config and the controller would be able to reconcile appropriately :
- non coco peer pods : always put `defaultInitdata` since policies are out of scope for this use case
- coco peer pods : set the user provided initdata if present or clear it out

snir911 committed:
Update updateConfigMap to support removing keys by value match

Modified the updateConfigMap function to accept a keysToRemove map
parameter that allows removing keys from the configmap when their
values match the specified values.

Signed-off-by: Snir Schreiber <ssheribe@redhat.com>

bpradipt commented on controllers/utils.go:
```diff
@@ -224,6 +224,15 @@ func updateConfigMap(client client.Client, logger logr.Logger, cmName string, na
+			update = true
+		}
+	}
+
```
What's the purpose of removing a key from existing configmap ?

snir911 replied:
See the next commit, this is needed in order to clear the permissive INITDATA in case it exist when CoCo fg is being enabled.
AFAIU enabling/disabling fg after installation is not supported, however, it's better to have this just to make sure we don't get unwanted pp configuration for CoCo

bpradipt replied:
Ok.

bpradipt commented:
@snir911 since this is still WIP, I was thinking if it makes sense to add the default registry cred to the initdata based on the cluster secret as part of this PR ?

Ref: https://github.com/confidential-containers/guest-components/blob/main/confidential-data-hub/example.config.toml#L121

gkurz commented on controllers/confidential_handler.go:
```diff
@@ -92,12 +97,19 @@ func (r *KataConfigOpenShiftReconciler) handleConfidentialPeerPods(state Feature
 	if state == Enabled {
 		peerpodsCMData = map[string]string{"DISABLECVM": "false"}
+		// Remove INITDATA if it matches the default value
+		keysToRemove = map[string]string{"INITDATA": defaultNonCCInitdata}
```
This will silently go against the intent of the user if they want to set such a permissive INITDATA on purpose... is it something we want ?

As said earlier, this is an ownership problem. It doesn't look right to tell the user they can set INITDATA if OSC is likely to overwrite it. It looks like user input should go to a dedicated CM or at least to a separate CUSTOM_INITDATA entry in the peerpods CM.

Cc @pmores for insights.

snir911 replied:
> This will silently go against the intent of the user if they want to set such a permissive INITDATA on purpose... is it something we want ? 

It's very unlikely to collide as CoCo users will normally be required to set in the initdata not only the agent policy but also aa/cdh toml, having said that, i can add a unique comment in the file "This initdata was created by the osc-operator" or something like that.

>  It doesn't look right to tell the user they can set INITDATA if OSC is likely to overwrite it.

Bare in mind we set it only for the non-CoCo users while initdata is officially supported only in CoCo. (FWIW, the concept of dummy (empty) initdata for pp is used also upstream).

At this point i find this solution fairly decent, however, I agree it's an ownership problem and open to better option if someone have it.

gkurz replied:
This ownership problem requires some redesign that we might not be able to perform this close to the release. Let's add a `This initdata was created by the osc-operator` marker inside the OSC generated one in the meantime then.

pmores replied:
In general I believe (as I said before) that every k8s resource should have a single owner.  It shouldn't be that different items of the same resource have different owners, let alone that that a single item has multiple owners, the reason being that it can get confusing and frustrating rather quickly.

In particular, I unfortunately don't have enough context to propose a specific solution here but I do believe that if we let users set something we should do our best to avoid overwriting it later.  I understand that the semantics of the current situation seemingly makes incidents unlikely.  I still feel there should be ways to do better - e.g. if I understand correctly the setting in question here only makes sense under certain circumstances.  This sounds like something to be enforced by a schema.

gkurz reviewed (approved):
/lgtm

Thanks @snir911 !
